File: 백준/Gold/1976. 여행 가자/여행 가자.py
n = int(input())
m = int(input())
graph = [list(map(int, input().split())) for _ in range(n)]
parents = list(range(n))
plan = list(map(int, input().split()))


def find(a):
    if parents[a] != a:
        parents[a] = find(parents[a])
    return parents[a]


def union(a, b):
    a = find(a)
    b = find(b)
    if a > b:
        parents[a] = b
    else:
        parents[b] = a


for i in range(n):
    for j in range(n):
        if graph[i][j] == 1:
            union(i, j)

result = "YES"
for i in range(1, m):
    if parents[plan[i] - 1] != parents[plan[0] - 1]:
        result = "NO"
        break

print(result)

File: 백준/Gold/1976. 여행 가자/여행 가자.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.StringTokenizer;

public class Main {

    static int[] parent;

    private static void union(int a, int b) {
        a = find(a);
        b = find(b);

        if (a > b) {
            parent[a] = b;
        } else {
            parent[b] = a;
        }
    }

    private static int find(int a) {
        if (parent[a] == a) {
            return a;
        }
        return parent[a] = find(parent[a]);
    }


    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int N = Integer.parseInt(br.readLine()); // 도시의 수 < 200
        int M = Integer.parseInt(br.readLine()); // 여행 계획에 속한 도시들의 수 < 1000

        parent = new int[N + 1];

        for (int i = 1; i < N + 1; i++) {
            parent[i] = i;
        }

        for (int i = 1; i < N + 1; i++) {

            int[] line = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

            for (int j = 0; j < N; j++) {
                if (line[j] == 1) {
                    union(i, j+1);
                }
            }
        }
        int[] plan = Arrays.stream(br.readLine().split(" ")).mapToInt(Integer::parseInt).toArray();

        int start = find(plan[0]);

        for (int now : plan) {
            if (start != find(now)) {
                System.out.println("NO");
                return;
            }
        }

        System.out.println("YES");

    }
}